benchmarks/inference-server/triton/README.md

```diff
@@ -0,0 +1,185 @@
+# AI on GKE: Benchmark TensorRT LLM on Triton Server
+
+This guide outlines the steps for deploying and benchmarking a TensorRT Large Language Model (LLM) on the Triton Inference Server within Google Kubernetes Engine (GKE). It includes the process of building a Docker container equipped with the TensorRT LLM engine and deploying this container to a GKE cluster.
+
+## Prerequisites
+
+Ensure you have the following prerequisites installed and set up:
+- Docker for containerization
+- Google Cloud SDK for interacting with Google Cloud services
+- Kubernetes CLI (kubectl) for managing Kubernetes clusters
+- A Hugging Face account to access models
+- NVIDIA GPU drivers and CUDA toolkit for building the TensorRT LLM Engine
+## Step 1: Build the TensorRT LLM Engine and Docker Image
+
+1. **Build the TensorRT LLM Engine:** Follow the instructions provided in the [TensorRT LLM backend repository](https://github.com/triton-inference-server/tensorrtllm_backend/blob/main/README.md) to build the TensorRT LLM Engine.
+
+2. **Upload the Model to Google Cloud Storage (GCS)**
+
+   Transfer your model engine to GCS using:
+   ```
+   gsutil cp -r your_model_folder gs://your_model_repo/all_models/ 
+   ```
+   *Ensure to replace your_model_repo with your actual GCS repository path.**
+   
+   **Alternate method: Add the Model repository and the relevant scripts to the image**
+
+   Construct a new image from Nvidia's base image and integrate the model repository and necessary scripts directly into it, bypassing the need for GCS during runtime. In the `tritonllm_backend` directory, create a Dockerfile with the following content:
+   ```Dockerfile
+   FROM nvcr.io/nvidia/tritonserver:23.10-trtllm-python-py3
+
+   # Copy the necessary files into the container
+   COPY all_models /all_models
+   COPY scripts /opt/scripts
+
+   # Install Python dependencies
+   RUN pip install sentencepiece protobuf
+
+   RUN chmod +x /opt/scripts/start_triton_servers.sh
+
+   CMD ["/opt/scripts/start_triton_servers.sh"]
+   ```
+
+    For the initialization script located at `/opt/scripts/start_triton_servers.sh`, follow the structure below:
+
+   ```start_triton_servers.sh
+   #!/bin/bash
+
+
+    # Use the environment variable to log in to Hugging Face CLI
+    huggingface-cli login --token $HUGGINGFACE_TOKEN
+
+
+
+
+    # Launch the servers (modify this depending on the number of GPU used and the exact path to the model repo)
+    /opt/tritonserver/bin/tritonserver \
```

Comment:
I'm wondering why we need to launch two different triton processes here. For multiple GPUs, just setting tensor_parallelism in the config should be enough right?

Reply:
actually, I rested without mpirun and it seems to set tensor_parallelism we need to run via mpirun. Otherwise this assertion fails: 
Internal: unexpected error when creating modelInstanceState: [TensorRT-LLM][ERRO |
|                |         | R] Assertion failed: mpiSize == tp * pp (/app/tensorrt_llm/cpp/tensorrt_llm/runtime/worldConf |
|                |         | ig.cpp:80) 
https://github.com/NVIDIA/TensorRT-LLM/blob/728cc0044bb76d1fafbcaa720c403e8de4f81906/cpp/tensorrt_llm/runtime/worldConfig.cpp#L78C1-L78C63
in the example above tp = 2 and pp = 1 and mpiSize should be 2* 1 = 2

Reply:
Interesting - my guess is that the triton configuration wasn't updated as expected.

Triton has a couple files for config it looks like: 
1. https://nvidia.github.io/TensorRT-LLM/_cpp_gen/runtime.html#gptjsonconfig-h
2. https://nvidia.github.io/TensorRT-LLM/_cpp_gen/runtime.html#worldconfig-h

This is throwing an error at the world config not having the expected number of gpu's for the way the model was built.

I think we need to pass through the gpu's somehow - eg. --gpus = all to indicate use all the existing GPUs for the model maybe?
https://github.com/triton-inference-server/tensorrtllm_backend?tab=readme-ov-file#option-1-launch-triton-server-within-triton-ngc-container

Or I also see this option here to pass through the world-size value which correlates to the GPUs to use: 
https://github.com/triton-inference-server/tensorrtllm_backend?tab=readme-ov-file#option-3-launch-triton-server-within-the-triton-container-built-via-docker

You'll see the model weights are configured to use tp=4 in the conversion in the related example: https://github.com/triton-inference-server/tensorrtllm_backend?tab=readme-ov-file#prepare-tensorrt-llm-engines

Reply:
Oh fun fact - that launch triton server script is backed by mpirun as well!
https://github.com/triton-inference-server/tensorrtllm_backend/blob/main/scripts/launch_triton_server.py#L67

I think Kaushik's observation is right that tensorrt_llm backend relies on mpirun for tensor parallelism.